azurelinuxagent/common/utils/processutil.py

```diff
@@ -0,0 +1,226 @@
+
+
+def _destroy_process(process, signal_to_send=signal.SIGKILL):
+    """
+    Completely destroy the target process. Close the stdout/stderr pipes, kill the process, reap the zombie.
+    If process is the leader of a process group, kill the entire process group.
+
+    :param Popen process: Process to be sent a signal
+    :param int signal_to_send: Signal number to be sent
+    """
+    process.stdout.close()
+    process.stderr.close()
+    try:
+        pid = process.pid
+        if os.getpgid(pid) == pid:
+            os.killpg(pid, signal_to_send)
+        else:
+            os.kill(pid, signal_to_send)
+        process.wait()
+    except OSError as e:
+        if e.errno != ESRCH:
+            raise
+        pass    # If the process is already gone, that's fine
+
+
+def capture_from_process_modern(process, cmd, timeout):
+    try:
+        stdout, stderr = process.communicate(timeout=timeout)
+    except subprocess.TimeoutExpired:
+        os.killpg(os.getpgid(process.pid), 9)
+        stdout, stderr = process.communicate()
+        msg = format_stdout_stderr(sanitize(stdout), sanitize(stderr))
+        raise ExtensionError("Timeout({0}): {1}\n{2}".format(timeout, cmd, msg))
+    except OSError as e:
+        _destroy_process(process, signal.SIGKILL)
+        raise ExtensionError("Error while running '{0}': {1}".format(cmd, e.strerror))
+    except ValueError:
+        _destroy_process(process, signal.SIGKILL)
+        raise ExtensionError("Invalid timeout ({0}) specified for '{1}'".format(timeout, cmd))
+    except Exception as e:
+        _destroy_process(process, signal.SIGKILL)
+        raise ExtensionError("Exception while running '{0}': {1}".format(cmd, e))
+
+    return stdout, stderr
+
+
+def capture_from_process_pre_33(process, cmd, timeout):
+    """
+    Can't use process.communicate(timeout=), so do it the hard way.
+    """
+    def kill_on_timeout(pid, watcher_timeout):
+        """
+        Check for the continued existence of pid once per second. If pid no longer exists, exit with code 0.
+        If timeout (in seconds) elapses, kill pid and exit with code 1.
+        """
+        for iteration in range(watcher_timeout):
+            time.sleep(1)
+            try:
+                os.kill(pid, 0)
+            except OSError as ex:
+                if 3 == ex.errno:   # Process no longer exists
+                    exit(0)
+        os.killpg(os.getpgid(pid), signal.SIGKILL)
+        exit(1)
+
+    watcher = Process(target=kill_on_timeout, args=(process.pid, timeout))
+    watcher.start()
+
+    try:
+        # Now, block "forever" waiting on process. If the timeout-limited Event wait in the watcher pops,
+        # it will kill the process and Popen.communicate() will return
+        stdout, stderr = process.communicate()
+    except OSError as e:
+        _destroy_process(process, signal.SIGKILL)
+        raise ExtensionError("Error while running '{0}': {1}".format(cmd, e.strerror))
```

Comment:
Why is this e.strerror and the others are just e?

Reply:
Everywhere I catch OSError, I explicitly display strerror, which is the string description of the underlying OS error (e.g. "File not found" or "No such process"). That string description is guaranteed to be present for OSError; it's actually the hallmark of that exception, which surfaces an errno from the underlying OS operations.

For all other exception types, the best I can do is convert the exception to a string and let the poor human find the pony under all the manure. For OSError, I can present a nice, clean pony.

Reply:
I wish there was a way to resolve a comment.